src/Sparrow/Allocators.Arena.cs

```diff
@@ -0,0 +1,332 @@
+            }                
+
+            // We will always cap the size of the Arena toward the MaxArenaSize.
+            if (preferredSize > allocator._options.MaxArenaSize)
+                preferredSize = allocator._options.MaxArenaSize;
+
+            allocator._preferredSize = preferredSize;
+            allocator._currentBuffer = new Pointer();
+            allocator._ptrStart = allocator._ptrCurrent = null;
+            allocator._used = 0;
+
+            allocator.TotalAllocated = 0;
+            allocator.Allocated = 0;
+            allocator.InUse = 0;
+        }
+
+        public void Renew(ref ArenaAllocator<TOptions> allocator)
+        {
+            if (!allocator._currentBuffer.IsValid)
+            {
+                allocator._currentBuffer = allocator._nativeAllocator.Allocate(ref allocator._nativeAllocator, (int)allocator._preferredSize);
+                allocator.Allocated += allocator._currentBuffer.Size;
+            }
+
+            if (allocator._olderBuffers != null)
+            {
+                // Free old buffers not being used anymore. A single arena ought to be enough for anybody, you know ;) 
+                foreach (var unusedBuffer in allocator._olderBuffers)
+                {
+                    _used += unusedBuffer.Size;
+
+                    Pointer ptr = unusedBuffer;
+
+                    allocator.Allocated -= ptr.Size;
+                    allocator._nativeAllocator.Release(ref allocator._nativeAllocator, ref ptr);
+                }
+
+                allocator._olderBuffers.Clear();
+            }
+
+            allocator._ptrStart = allocator._ptrCurrent = (byte*)allocator._currentBuffer.Ptr;
+            allocator._used = 0;
+            allocator.InUse = 0;
+        }
+
+        public void OnAllocate(ref ArenaAllocator<TOptions> allocator, Pointer ptr)
+        {
+            // This allocator does not keep track of anything.
+        }
+
+        public void OnRelease(ref ArenaAllocator<TOptions> allocator, Pointer ptr)
+        {
+            // This allocator does not keep track of anything.
+        }
+
+        public void Dispose(ref ArenaAllocator<TOptions> allocator)
+        {            
+            if (allocator._olderBuffers != null)
+            {
+                // Free the remanent buffers on local storage from last cycle regrowths
+                foreach (var unusedBuffer in allocator._olderBuffers)
+                {
+                    _used += unusedBuffer.Size;
+
+                    Pointer ptr = unusedBuffer;
+                    allocator._nativeAllocator.Release(ref _nativeAllocator, ref ptr);
+                }
+            }
+
+            if (allocator._currentBuffer.IsValid)
+            {
+                allocator._nativeAllocator.Release(ref _nativeAllocator, ref _currentBuffer);
+            }
+
+            allocator._nativeAllocator.Dispose(ref allocator._nativeAllocator);
```

Comment:
Why do we pass the caller (`_nativeAllocator`) here and when calling `Release()` a few lines up?

Reply:
That is a design decision that may or may not have merits, but until I have something working really well I won't revisit. The idea of why allocators go always by reference is because C# does not support static methods on generic types. The problem is that you can create copies of the allocators by mistake, that prevents that and also helps the JIT to figure out some stuff when inlining this kind of code at the caller size ``Allocator<T>``. It's highly probable that we may get away without it, but because of how difficult this code is I preferred to err on the side of safety.